pyccel/ast/core.py

```diff
@@ -1313,6 +1329,118 @@ def declarations(self):
         return [Declare(i.dtype, i) for i in self.variables]
 
 
+#==============================================================================
+class Iterable(Basic):
+    """
+    Wrapper around iterable types helping to convert between those
+    types and a range (necessary in low level languages, e.g. C and Fortran)
+
+    Paramaters
+    ----------
+    iterable : acceptable_iterator_type
+                The iterator being wrapped
+    """
+    indexless_types = (PythonRange, PythonEnumerate)
+    acceptable_iterator_types = (Variable, PythonMap, PythonZip, PythonEnumerate, Product, PythonRange)
+    __slots__ = ('_iterable','_indices','_num_indices_required')
+    _attribute_nodes = ('_iterable','_indices')
+
+    def __init__(self, iterable):
+        self._iterable = iterable
+        self._indices  = None
+
+        if isinstance(iterable, self.indexless_types):
+            self._num_indices_required =  0
+        elif isinstance(iterable, Product):
+            self._num_indices_required = len(iterable.elements)
+        elif isinstance(iterable, self.acceptable_iterator_types):
+            self._num_indices_required = 1
+        else:
+            raise TypeError("Unknown iterator type")
+
+        super().__init__()
+
+    @property
+    def num_indices_required(self):
+        """ Number of indices which must be generated in order to
+        convert this iteratable to a range
+        """
+        return self._num_indices_required
+
+    def set_indices(self, *indices):
+        """ Set the indices used to define the iterator/target from the range
+        """
+        self._indices = indices
```

Comment:
Hmm, this is quite hard to describe. As it says in the class docstring an `Iterable` is a wrapper around iterable types which helps to convert between those types and a range for languages where this is necessary. So for example:
```
for ai in a:
```
Should be translated to:
```
for i in range(len(a)):
    ai = a[i]
```
The iterator `i` here must be created by the semantic stage (which has knowledge about potential name clashes) and then passed to `Iterable` so it can be used when generating the range target (`a[i]`). This function passes those indices (`i`)